File: pack/Context.java
package common.pack;

import common.io.PackLoader.ZipDesc.FileDesc;
import common.system.P;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public interface Context {

	enum ErrType {
		FATAL, ERROR, WARN, INFO, NEW, CORRUPT, DEBUG
	}

	interface RunExc {

		void run() throws Exception;

	}

	interface SupExc<T> {

		T get() throws Exception;

	}

	static void check(boolean bool, String str, File f) throws IOException {
		if (!bool)
			throw new IOException("failed to " + str + " file " + f);
	}

	static void check(File f) throws IOException {
		if (!f.getParentFile().exists())
			check(f.getParentFile().mkdirs(), "create", f);
		if (!f.exists())
			check(f.createNewFile(), "create", f);
	}

	static void delete(File f) throws IOException {
		if (f == null || !f.exists())
			return;
		if (f.isDirectory())
			for (File i : f.listFiles())
				delete(i);
		check(f.delete(), "delete", f);
	}

	static void renameTo(File a, File b) {
		if (!b.getParentFile().exists())
			b.getParentFile().mkdirs();
		if (b.exists())
			b.delete();
		a.renameTo(b);
	}

	boolean confirmDelete();

	File getAssetFile(String string);

	File getAuxFile(String string);

	InputStream getLangFile(String file);

	File getUserFile(String string);

	File getWorkspaceFile(String relativePath);

	void initProfile();

	default boolean noticeErr(Context.RunExc r, ErrType t, String str) {
		try {
			r.run();
			return true;
		} catch (Exception e) {
			noticeErr(e, t, str);
			return false;
		}
	}

	default <T> T noticeErr(Context.SupExc<T> r, ErrType t, String str) {
		try {
			return r.get();
		} catch (Exception e) {
			noticeErr(e, t, str);
			return null;
		}
	}

	default <T> T noticeErr(Context.SupExc<T> r, ErrType t, String str, Runnable onFail) {
		try {
			return r.get();
		} catch (Exception e) {
			onFail.run();
			noticeErr(e, t, str);
			return null;
		}
	}

	void noticeErr(Exception e, ErrType t, String str);

	boolean preload(FileDesc desc);

	void printErr(ErrType t, String str);

}

File: pack/Context.kt
package common.pack

import common.io.PackLoader.ZipDesc.FileDesc
import common.pack.Context
import java.io.File
import java.io.IOException

interface Context {

    enum class ErrType {
        FATAL, ERROR, WARN, INFO, NEW, CORRUPT
    }

    fun interface RunExc {
        @Throws(Exception::class)
        fun run()
    }

    fun interface SupExc<T> {
        @Throws(Exception::class)
        fun get(): T
    }

    fun confirmDelete(): Boolean
    fun getAssetFile(string: String): File
    fun getLangFile(file: String): File
    fun getPackFolder(): File
    fun getUserFile(string: String): File
    fun getWorkspaceFile(relativePath: String): File
    fun initProfile()
    fun noticeErr(r: Context.RunExc, t: ErrType, str: String?): Boolean {
        return try {
            r.run()
            true
        } catch (e: Exception) {
            noticeErr(e, t, str)
            false
        }
    }

    fun <T> noticeErr(r: SupExc<T>, t: ErrType, str: String?): T? {
        return try {
            r.get()
        } catch (e: Exception) {
            noticeErr(e, t, str)
            null
        }
    }

    fun noticeErr(e: Exception, t: ErrType, str: String?)
    fun preload(desc: FileDesc): Boolean
    fun printErr(t: ErrType, str: String?)

    companion object {
        @Throws(IOException::class)
        fun check(bool: Boolean, str: String, f: File) {
            if (!bool) throw IOException("failed to $str file $f")
        }

        @Throws(IOException::class)
        fun check(f: File) {
            if (!f.parentFile.exists()) Context.Companion.check(f.parentFile.mkdirs(), "create", f)
            if (!f.exists()) Context.Companion.check(f.createNewFile(), "create", f)
        }

        @Throws(IOException::class)
        fun delete(f: File?) {
            if (f == null || !f.exists()) return
            if (f.isDirectory) for (i in f.listFiles()) Context.Companion.delete(i)
            Context.Companion.check(f.delete(), "delete", f)
        }
    }
}
